### PR TITLE
untangeling datatypes

### PR DESCRIPTION
**Pull request by i-walker**
finishing the arrow-core-data list from [here](https://github.com/arrow-kt/arrow/pull/1481#issue-287079927)
- [x] SequenceK -> arrow-core-data
- [x] SortedMap -> arrow-core-data
- [x] SortedMapK -> arrow-core-data
- [x] create mtl-data, mtl and ui-data modules
- [x] OptionT -> arrow-mtl-data
- [x] Reader -> arrow-mtl-data
- [x] State -> arrow-mtl-data
- [x] StateT -> arrow-mtl-data
- [x] Cokleisli -> arrow-mtl-data
- [x] Coreader -> -> arrow-mtl-data
- [x] EitherT -> arrow-mtl-data
- [x] Kleisli -> arrow-mtl-data
- [x] move Tests from arrow-extras to their respective modules
- [x] WriterT -> arrow-mtl-data
- [x] Day -> arrow-ui-data
- [x] Moore -> arrow-ui-data
- [x] Store -> arrow-ui-data
- [x] Sum -> arrow-ui-data
- [x] remove Coproduct
- [x] remove arrow-extras modules

#### The following lower-priority PR includes:
```
- [ ] imports added and playground also instead arrow:ank 
- [ ] move the url to the apidocs permalink look up in the webpage
aka /docs/apidocs/arrow-extras/arrow.data/-non-empty-list/
== builds run the site == 
 ./gradlew clean dokka :arrow-docs:runAnk ; cd modules/docs/build/site; jekyll serve;

- [ ] Edit permalinks in sidebar.yml for all moved Datatypes and move them into the correct subdir url
- [ ] Edit Doc's for arrow-mtl + arrow-core + (Introduce) arrow-ui +  
- [ ] move Documentation to KDocs (depending on the weight of the tasks I may split this into another PR)
```
#### For Future Reference
```
- [ ] * issue*: remove mtl-data from arrow-test deps // let it -> arrow-test will be removed 
- [ ] TODO: Add title and Url for arrow-core, arrow-ui, arrow-mtl-data at sidebar.yaml
- [ ]  TODO: change readme for quickstart/libraries ReadMe
- [ ]  Todo remove arrow-syntax
```
#### high priority PR for  fx: 
```
- [ ] arrow-effects -> arrow-fx 
refactor effects -> fx
- [ ] new structure
arrow-fx
arrow-fx-rx2
arrow-fx-reactor
- [ ] clean up HKT in effects
-> delete @higherkind in effects 
on top of the class
```

### CODE REVIEW

**Comment by i-walker** [comment]
Let me finish the list and then we can merge

File: README.md [hunk 1]
```diff
@@ -72,7 +72,7 @@ Recomended for most use cases:
 Trimmed down versions:
 
 - `arrow-core-data` (Only data types)
-- `arrow-core-extensions` (Only type class extensions)
+- `arrow-core` (Only type class extensions)
```
Comment (pakoito) [inline README.md]: Data types and class extensions

File: modules/aql/arrow-query-language/build.gradle [hunk 2]
```diff
@@ -10,6 +10,7 @@ dependencies {
     compileOnly project(':arrow-meta')
     testCompileOnly project(':arrow-meta')
     testRuntime("org.junit.vintage:junit-vintage-engine:$jUnitVintageVersion")
+
```
Comment (pakoito) [inline modules/aql/arrow-query-language/build.gradle]: -

**Comment by raulraja** [comment]
@i-walker heads up you may want to fix the conflicts before continuing

File: README.md [hunk 3]
```diff
@@ -90,16 +90,15 @@ Add the dependencies into the project's `build.gradle`
...
     compile "io.arrow-kt:arrow-syntax:$arrow_version"
     compile "io.arrow-kt:arrow-typeclasses:$arrow_version"
-    compile "io.arrow-kt:arrow-extras-data:$arrow_version"
-    compile "io.arrow-kt:arrow-extras-extensions:$arrow_version"
     kapt    "io.arrow-kt:arrow-meta:$arrow_version"
     
     compile "io.arrow-kt:arrow-query-language:$arrow_version" //optional
     compile "io.arrow-kt:arrow-free-data:$arrow_version" //optional
     compile "io.arrow-kt:arrow-free-extensions:$arrow_version" //optional
+    compile "io.arrow-kt:arrow-mtl-data:$arrow_version" //optional
```
Comment (pakoito) [inline README.md]: Given that mtl now includes data too, I'd say remove the half-versions from here.

File: README.md [hunk 4]
```diff
@@ -90,16 +90,15 @@ Add the dependencies into the project's `build.gradle`
 def arrow_version = "0.9.1-SNAPSHOT"
 dependencies {
     compile "io.arrow-kt:arrow-core-data:$arrow_version"
```
Comment (pakoito) [inline README.md]: Given that core now includes core-data too, I'd say remove the half-versions from here.

File: modules/core/arrow-core-data/src/main/kotlin/arrow/core/NonFatal.kt [hunk 5]
```diff
@@ -54,8 +53,7 @@ fun NonFatal(t: Throwable): Boolean =
  *
  * ```kotlin:ank:playground
  * import arrow.*
- * import arrow.core.*
- * import arrow.data.*
+ * import arrow.core.*]
```
Comment (pakoito) [inline modules/core/arrow-core-data/src/main/kotlin/arrow/core/NonFatal.kt]: That extra bracket?

File: modules/core/arrow-core-data/src/test/kotlin/arrow/extensions/ComposedInstancesTest.kt [hunk 6]
```diff
@@ -130,7 +106,6 @@ class ComposedInstancesTest : UnitSpec() {
 
     testLaws(
       FunctorLaws.laws(ComposedFunctor(Option.functor(), NonEmptyList.functor()), cf, EQ_OPTION_NEL),
-      FunctorFilterLaws.laws(ComposedFunctorFilter(OptionT.functorFilter(Id.monad()), OptionT.functorFilter(NonEmptyList.monad())), { OptionT.just(Id.monad(), OptionT.just(NonEmptyList.monad(), it)).nest() }, EQ_OPTIONT_ID_NEL),
```
Comment (pakoito) [inline modules/core/arrow-core-data/src/test/kotlin/arrow/extensions/ComposedInstancesTest.kt]: What happened here? Why was this deleted?

Comment (i-walker) [reply modules/core/arrow-core-data/src/test/kotlin/arrow/extensions/ComposedInstancesTest.kt]: It needed to move to mtl. We had to seperate the Transformer Tests from the Core one

File: modules/core/arrow-core/gradle.properties [hunk 7]
```diff
@@ -1,4 +1,4 @@
 # Maven publishing configuration
 POM_NAME=Arrow Core Extensions
-POM_ARTIFACT_ID=arrow-core-extensions
```
Comment (pakoito) [inline modules/core/arrow-core/gradle.properties]: @raulraja @JorgeCastilloPrz didn't we agree that there'd be arrow-core-data, arrow-core-extensions; and a separate package arrow-core that'd only reexport both?

File: modules/docs/arrow-docs/docs/_data/sidebar.yml [hunk 8]
```diff
@@ -35,18 +35,11 @@ options:
       - title: Type Classes
         url: /docs/apidocs/arrow-typeclasses/
 
+      # TODO: Add title and Url for arrow-core, arrow-ui, arrow-mtl-data
```
Comment (pakoito) [inline modules/docs/arrow-docs/docs/_data/sidebar.yml]: This needs to be completed for the PR to be considered done.

File: modules/docs/arrow-docs/docs/docs/quickstart/libraries/README.md [hunk 9]
```diff
@@ -49,26 +49,26 @@ Datatypes: [`Const`]({{ '/docs/typeclasses/const/' | relative_url }})
...
+Dependencies: arrow-core-data
+// TODO: Change locations in Readme before merge
```
Comment (pakoito) [inline modules/docs/arrow-docs/docs/docs/quickstart/libraries/README.md]: FIX before merge!

File: modules/docs/arrow-docs/docs/docs/quickstart/libraries/README.md [hunk 10]
```diff
@@ -49,26 +49,26 @@ Datatypes: [`Const`]({{ '/docs/typeclasses/const/' | relative_url }})
...
+Datatypes: [`Cokleisli`]({{ '/docs/datatypes/cokleisli/' | relative_url }}), [`Coreader`]({{ '/docs/datatypes/coreader/' | relative_url }}), [`Ior`]({{ '/docs/arrow/data/ior/' | relative_url }}), [`ListK`]({{ '/docs/arrow/data/listk/' | relative_url }}), [`NonEmptyList`]({{ '/docs/arrow/data/nonemptylist/' | relative_url }}), [`SequenceK`]({{ '/docs/arrow/data/sequencek/' | relative_url }}), [`SortedMapK`]({{ '/docs/arrow/data/sortedmapk/' | relative_url }}), [`StateT`]({{ '/docs/arrow/data/statet/' | relative_url }}), [`WriterT`]({{ '/docs/arrow/data/writert/' | relative_url }}), [`EitherT`]({{ '/docs/arrow/data/eithert/' | relative_url }}), [`Kleisli`]({{ '/docs/arrow/data/kleisli/' | relative_url }}), [`MapK`]({{ '/docs/arrow/data/mapk/' | relative_url }}), [`OptionT`]({{ '/docs/arrow/data/optiont/' | relative_url }}), [`Reader`]({{ '/docs/arrow/data/reader/' | relative_url }}), [`SetK`]({{ '/docs/arrow/data/setk/' | relative_url }}), [`State`]({{ '/docs/arrow/data/state/' | relative_url }}), [`Validated`]({{ '/docs/arrow/data/validated/' | relative_url }})
 
-### arrow-(core, extras, effects)-extensions
+### arrow-core, arrow-effects-extensions
```
Comment (pakoito) [inline modules/docs/arrow-docs/docs/docs/quickstart/libraries/README.md]: This seems incorrect. We should keep arrow-core-extensions don't we?

What we agreed is that the extensions will include _all of them_, also the MTL ones.

File: modules/mtl/arrow-mtl-data/src/test/kotlin/arrow/mtl/OptionTTest.kt [hunk 11]
```diff
@@ -71,6 +93,12 @@ class OptionTTest : UnitSpec() {
         OptionT.applicative(Option.monad()),
         EQ()),
 
+      FunctorFilterLaws.laws(
+        ComposedFunctorFilter(OptionT.functorFilter(Id.monad()),
```
Comment (pakoito) [inline modules/mtl/arrow-mtl-data/src/test/kotlin/arrow/mtl/OptionTTest.kt]: This tests the ComposedFunctorFilter rather than OptionT, right?

File: modules/core/arrow-core-data/src/main/kotlin/arrow/core/NonFatal.kt (hunk 5 above)
Comment (i-walker) [inline modules/core/arrow-core-data/src/main/kotlin/arrow/core/NonFatal.kt]:
```suggestion
 * import arrow.core.*
```

File: modules/core/arrow-core/gradle.properties (hunk 7 above)
Comment (i-walker) [reply modules/core/arrow-core/gradle.properties]: That changed, because of unnecessary overhead. 
Now there is: 
- arrow-core-data
- arrow-core 

Same in mtl:
- arrow-mtl-data
- arrow-mtl

@raulraja @JorgeCastilloPrz should we do the same for free, effects. and other modules ?

File: modules/docs/arrow-docs/docs/_data/sidebar.yml (hunk 8 above)
Comment (i-walker) [reply modules/docs/arrow-docs/docs/_data/sidebar.yml]: I open for discussion

Comment (i-walker) [reply modules/docs/arrow-docs/docs/_data/sidebar.yml]: We moved this to an issue: 
I will create it and write down steps, maybe other contributors can help me and We can review it step by step

File: modules/docs/arrow-docs/docs/docs/quickstart/libraries/README.md (hunk 9 above)
Comment (i-walker) [reply modules/docs/arrow-docs/docs/docs/quickstart/libraries/README.md]: The relative links will be changed anyways so I will create the issue push another commit without the Todo's and we're good

File: modules/core/arrow-core-data/build.gradle [hunk 12]
```diff
@@ -10,9 +10,8 @@ dependencies {
     testRuntime("org.junit.vintage:junit-vintage-engine:$jUnitVintageVersion")
     testCompile "io.kotlintest:kotlintest-runner-junit4:$kotlinTestVersion"
     testCompile project(':arrow-test')
-    testCompile project(':arrow-syntax')
+    testCompile project(':arrow-syntax') // TODO: remove
```
Comment (pakoito) [inline modules/core/arrow-core-data/build.gradle]: What is the reason behind this? If this remains here, it'll bloat the binary by almost a megabyte.

Comment (pakoito) [inline modules/core/arrow-core-data/build.gradle]: Why remove it? It's only for testing
